Websocket sansio implementataion

gourav-kandoria opened the pull request:
This PR provides the implementation of websocket protocol using websockets sansI/O layer.

task details - https://github.com/encode/uvicorn/issues/1908


## Checklist for @Kludex
- [ ] `data_receive` shouldn't be using `Exception`. What is the right exception there?
- [X] Check if `max_size` is being tested, and if not, test it. EDIT: I've checked it - it's not being tested.
- [ ] Add `WebSocketsSansIOProtocol` test for `max_size`.
- [x] Should we remove `CONT` and `PONG` related code?
- [ ] Add documentation.
- [x] Increase the coverage number.
- [x] Add `--ws websockets-sansio`.
- [ ] Check if https://github.com/encode/uvicorn/pull/2060#discussion_r1277501171 is true.
- [ ] Answer the following questions:
  - [ ] Should we deprecate `websockets` legacy version?
  - [ ] Should we add a `UserWarning` saying that it's experimental, and asking users to comment in a discussion X?

gourav-kandoria committed:
Implement websockets_sans_impl.py

gourav-kandoria committed:
Implement websockets_sans_impl.py

gourav-kandoria committed:
add surrogate errors in decode

gourav-kandoria committed:
add surrogate errors in decode

gourav-kandoria committed:
fix lint issues

gourav-kandoria committed:
fix lint issues

gourav-kandoria committed:
fix mypy failing issues

gourav-kandoria committed:
fix mypy failing issues

gourav-kandoria committed:
fix lint issues

gourav-kandoria committed:
fix lint issues

gourav-kandoria committed:
fix typing issues

gourav-kandoria committed:
fix typing issues

gourav-kandoria committed:
Fix extension tests failing

gourav-kandoria committed:
Fix extension tests failing

gourav-kandoria committed:
Fix extension tests failing

gourav-kandoria committed:
Fix extension tests failing

Kludex commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,392 @@
+    from typing import Literal
+
+if typing.TYPE_CHECKING:
+    from asgiref.typing import (
```
We are using uvicorn._types instead.

gourav-kandoria replied:
@Kludex sure will change this

Kludex reviewed (commented):
I'll review on the weekend.

Kludex reviewed (commented):
I'm going to explain +- how is going to be my review process, so you can try to speed it up:

1. I'll run the test suite, and check the coverage, and then I'll try to understand why some lines are not covered.
2. I'll check the type hints, and make sure that they make sense. 
3. There are some tests that only work for `wsproto` or for `websockets`. I'm going to make sure that we are also running this protocol implementation where it makes sense.

valentin994 commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,392 @@
+
+        # Buffers
+        self.bytes: "bytes" = b""
+        print(len(self.tasks))
```
I guess you wouldn't leave this in for master

valentin994 commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,392 @@
+
+    def connection_lost(self, exc: typing.Optional[Exception]) -> None:
+        self.connections.remove(self)
+        print("came in connection lost : ", exc)
```
leftover print?

gourav-kandoria replied:
@valentin994 yes, would remove these types of statement. For now opened the pr for just review and suggestions

valentin994 commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,392 @@
+        if self.handshake_initiated and not self.close_sent:
+            self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
+
+    def data_received(self, data: bytes) -> None:
```
https://websockets.readthedocs.io/en/stable/howto/sansio.html
if I'm following this right after receiving data you should call receive_eof()

gourav-kandoria replied:
@valentin994 @Kludex we don't have to call receive_eof() after receiving data. Because,  protocol ensures that (https://docs.python.org/3/library/asyncio-protocol.html#asyncio.Protocol.data_received) data_received will only be called only when the data is there. In case when eof is received from client side, the transport will close and connection_lost will be called.

gourav-kandoria committed:
correct types import

gourav-kandoria committed:
correct types import

gourav-kandoria committed:
correct types import and mypy issues

gourav-kandoria committed:
correct types import and mypy issues

gourav-kandoria committed:
fix typo

gourav-kandoria committed:
fix typo

gourav-kandoria committed:
Merge remote-tracking branch 'upstream/master' into websocket-sansio-implementataion

gourav-kandoria committed:
Replace ServerConnection with ServerProtocol due to upgradation of websockets version

gourav-kandoria committed:
Replace ServerConnection with ServerProtocol due to upgradation of websockets version

gourav-kandoria committed:
Merge remote-tracking branch 'upstream/master' into websocket-sansio-implementataion

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+        _loop: typing.Optional[asyncio.AbstractEventLoop] = None,
+    ) -> None:
+        if not config.loaded:
+            config.load()
```
in the test_websockets . this line is not being covered for any implementation of websockets.
as an http upgrade is happening there and httpProtocol classes are passing already loaded config in that cases.
upon inspection, I found that this line was covered for websockets_impl.py due test_websocket_auto.py

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+            self.conn.receive_data(data)
+        except Exception:
+            self.logger.exception("Exception in ASGI server")
+            self.transport.close()
```
This lines(117-119) are not being covered. Because, I guess in the test suite there not any test which after connection establishment sends data which causes receive_data to throw exception such as data which don't follow websocket spec.

Kludex replied:
We still need to test it. It's unlikely that we want `Exception` there as well.

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+                self.handle_connect(event)
+            if isinstance(event, Frame):
+                if event.opcode == websockets.frames.Opcode.CONT:
+                    self.handle_cont(event)
```
This line is not covered. as in test suite there is not any test which is sending data in multiple frames. So the whole data is coming in the first frame itself. with fin also set to true. 
Although I had tested it locally on my system.
where I sent multiple frames. in that case, the sequence of frames would be
websockets.frames.Opcode.TEXT or websockets.frames.Opcode.BINARY
followed by ... websockets.frames.Opcode.CONT
then may be more websockets.frames.Opcode.CONT frames
and finally in last websockets.frames.Opcode.CONT frame, fin would be set to true

Kludex replied:
Is this related to HTTP/2?

gourav-kandoria replied:
@Kludex No, In case HTTP/1, according to websocket spec, this could be the behaviour

Kludex replied:
got it

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+    def handle_cont(self, event: Frame) -> None:
+        self.bytes += event.data
+        if event.fin:
+            self.send_receive_event_to_app()
```
These lines(195-197) are not covered due to the same reason as I mentioned for line no. 141

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+                elif event.opcode == websockets.frames.Opcode.PING:
+                    self.handle_ping(event)
+                elif event.opcode == websockets.frames.Opcode.PONG:
+                    self.handle_pong(event)
```
as per this document - https://datatracker.ietf.org/doc/html/rfc6455#autoid-29
[5.5.3] Pong
 The Pong frame contains an opcode of 0xA.
   A Pong frame sent in response to a Ping frame must have identical
   "Application data" as found in the message body of the Ping frame
   being replied to.
   If an endpoint receives a Ping frame and has not yet sent Pong
   frame(s) in response to previous Ping frame(s), the endpoint MAY
   elect to send a Pong frame for only the most recently processed Ping
   frame.
   A Pong frame MAY be sent unsolicited.  This serves as a
   unidirectional heartbeat.  A response to an unsolicited Pong frame is
   not expected.
   
   So, the websockets sansI/O layer chooses not to propagate the received PONG frame event.
   That's why It is not being covered.

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+        self.transport.writelines(output)
+
+    def handle_pong(self, event: Frame) -> None:
+        pass
```
same reason as mentioned for line number 149

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,389 @@
+from uvicorn.server import ServerState
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal
```
@Kludex I guess, this condition can be removed as uvicorn requirement for python is atleast 3.8. right?

Kludex replied:
yes

Kludex committed:
Remove conditional on imports

Kludex committed:
Remove conditional on imports

Kludex committed:
Fix typos, and small details

Kludex committed:
Fix typos, and small details

Kludex committed:
Refactor small things

Kludex committed:
Refactor small things

Kludex committed:
Fix linter

Kludex committed:
Fix linter

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,378 @@
+                elif event.opcode == Opcode.PONG:
+                    self.handle_pong(event)
+                elif event.opcode == Opcode.CLOSE:
+                    self.handle_close(event)
```
@Kludex @aaugustin Could you please help with this?
one scenario, where we can expect close frame is when the client sends it.
What other scenarios are there where we can expect it?
Can we expect it in scenarios where the we received unexpected data
and how can we have access to the close code, so that same code can be passed in disconnect event to asgi app

gourav-kandoria commented on uvicorn/protocols/websockets/websockets_sansio_impl.py:
```diff
@@ -0,0 +1,378 @@
+    def data_received(self, data: bytes) -> None:
+        try:
+            self.conn.receive_data(data)
+        except Exception:
```
@Kludex I guess, the "except Exception" block can be removed from here.  Upon reading code of Protocol, It seems that exception thrown by code are handled there internally and the connection is being made to fail. @aaugustin could you please confirm , if this is the case

Kludex replied:
No, that's not the case... Since you are using `asyncio.Protocol`.

gourav-kandoria replied:
@Kludex oh sorry for miscommunication, here I meant websockets Protocol (the library which we are using)

gourav-kandoria committed:
Merge branch 'websocket-sansio-implementataion' of https://github.com/gourav-kandoria/uvicorn into websocket-sansio-implementataion

gourav-kandoria committed:
Merge branch 'master' of https://github.com/encode/uvicorn into websocket-sansio-implementataion

gourav-kandoria committed:
Add tests for websocket server for receiving multiple frames

gourav-kandoria committed:
Add tests for websocket server for receiving multiple frames

gourav-kandoria committed:
Remove checking of PONG event after receiving data

As it won't be propagated as an event by the websockets sansIO protocol

gourav-kandoria committed:
Remove checking of PONG event after receiving data

As it won't be propagated as an event by the websockets sansIO protocol

gourav-kandoria committed:
Revert "Remove checking of PONG event after receiving data"

This reverts commit 808f9515147d8b40b70fbd7761df097a8540e1a9.

gourav-kandoria committed:
Revert "Remove checking of PONG event after receiving data"

This reverts commit 808f9515147d8b40b70fbd7761df097a8540e1a9.

gourav-kandoria committed:
"Remove checking of PONG event after receiving data"

As websockets sansio protocol is not propagating this event after receiving data

gourav-kandoria committed:
"Remove checking of PONG event after receiving data"

As websockets sansio protocol is not propagating this event after receiving data

Kludex committed:
Create WSType on the test suite

Kludex committed:
Create WSType on the test suite

Kludex committed:
Add WebSocketsSansIOProtocol to the CLI

Kludex committed:
Add WebSocketsSansIOProtocol to the CLI

Kludex reviewed (changes requested):
`max_size` is not being tested on `WebSocketSansIOProtocol` - There's a test that uses `max_size` on the `test_websockets.py`.

The `Exception` part, we need to add the right exception, and test it.

gourav-kandoria committed:
Make changes for testing payload max_size limit

gourav-kandoria committed:
Make changes for testing payload max_size limit

gourav-kandoria committed:
Make changes for testing payload max_size limit

gourav-kandoria committed:
Make changes for testing payload max_size limit

gourav-kandoria committed:
fix lint issue

gourav-kandoria committed:
fix lint issue

gourav-kandoria committed:
increase msg size from 11 to 32

gourav-kandoria committed:
increase msg size from 11 to 32

gourav-kandoria commented on tests/protocols/test_websocket.py:
```diff
@@ -968,7 +968,7 @@ async def get_subprotocol(url: str):
         (MAX_WS_BYTES_PLUS1, MAX_WS_BYTES, 1009),
         (10, 10, 0),
-        (11, 10, 1009),
+        (32, 10, 1009),
```
@Kludex I am facing strange situation, in this test case on keeping the message size greater than 31 . Test is passing, but on values from 11 to 31 , it is failing. not sure, where there is some issue with my code, or ServerProtocol or with the client. Could you please assist in this

Kludex replied:
I can, but it will take some weeks. I'm sorry. I'm a bit busy.

gourav-kandoria replied:
sure, whenever the time permits you :)

gourav-kandoria replied:
Hi, actually I was able to find the reason for it. 
In the test just like, There is max_size limit on the server. There were also max_size limit on the client side also
like this

```
async with websockets.client.connect(
            url, max_size=client_size_sent
   ) as websocket:
```
   
   So, in case when client_size_sent is 11, The close frame is being received with code 1009, along with a reason as "over size limit ({length} > {max_size} bytes)". Now the total size of this close which include the close code and the reason is summing more than 11. So, before the client processes this close frame, it is detecting the payload is bigger than 11 bytes. which is causing the client to close the connection on its end without processing the close frame from the server.
   
 So, for this test to work correctly, I have changed the client initiation logic from 
```
        async with websockets.client.connect(
            url, max_size=client_size_sent
        ) as websocket:
```

to this on latest change

`        async with websockets.client.connect(url, max_size=MAX_WS_BYTES) as websocket:`

Because, anyway, we want to test that atleast server is able send the close frame correctly and don't care what happens at client end

gourav-kandoria committed:
increase client max_limit

gourav-kandoria committed:
increase client max_limit

gourav-kandoria committed:
Empty-Commit-to-trigger-pipeline

gourav-kandoria committed:
Empty-Commit-to-trigger-pipeline

Kludex committed:
Use WSProtocolType

Kludex committed:
Use future annotations on websocket sansio implementation

Kludex committed:
WIP websockets denial response extension

gourav-kandoria committed:
Merge branch 'websocket-sansio-implementataion' of https://github.com/gourav-kandoria/uvicorn into websocket-sansio-implementataion
